### PR TITLE
Feature: Quiver Display

### PR DESCRIPTION
**Pull request by ItsEmpa**
## Dependencies
- #1202 
- #1209 

## What
Adds a Quiver Display that shows the amount of arrows you have, an alert to get after a dungeon/kuudra run if you ran out/are low on certain arrows, and a customizable amount to get that reminder and a low quiver notification. Also makes QuiverAPI use the item in the 9th slot of the hotbar to get more accurate numbers.

<details>
<summary>Images</summary>

image

</details>

<details>
<summary>TODO</summary>

- [x] Create QuiverUpdateEvent and use to update display.
- [x] Option to only show Quiver Display with bow in inventory.
- [x] Hide arrow count when wearing Skeleton Master Chestplate.
- [x] Remind at end of Dungeon run if low quiver.
- [x] Create KuudraEnterEvent and KuudraCompleteEvent.
- [x] Fix display not appearing on startup until QuiverUpdateEvent.
- [x] Refactor Low Quiver Alert inside Quiver Display and add a customizable amount at which to alert.
- [x] Remind at end of Kuudra run if low quiver.
</details>

## Changelog New Features
+ Added Quiver Display. - Empa
    * Only shows the type of arrow when wearing a Skeleton Master Chestplate.
+ Added Low Quiver Reminder at the end of Dungeon/Kuudra run. - Empa

## Changelog Improvements
+ Added customizable amount at which to get a Low Quiver Notification / Quiver Reminder. - Empa

## Changelog Technical Details
+ Added slot to OwnInventoryItemUpdateEvent. - Empa
+ Added QuiverUpdateEvent. - Empa
+ Use Quiver item in 9th hotbar slot for QuiverAPI. - Empa

## Changelog Removed Features
+ Removed Bow Sound distance setting. - Empa

### CODE REVIEW

**Comment by ItsEmpa** [comment]
Finished what i wanted to do with this (mostly), only thing left in my todo is merging the dungeon and kuudra enter/complete events into an instance event, but apart from that it should be ready for review.

**Comment by ItsEmpa** [comment]
The description and names of the new configs also still have to be improved

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/GUIConfig.java [hunk 1]
```diff
@@ -64,6 +64,11 @@ public class GUIConfig {
     @Accordion
     public ModifyWordsConfig modifyWords = new ModifyWordsConfig();
 
+    @Expose
+    @ConfigOption(name = "Quiver Display", desc = "")
+    @Accordion
+    public QuiverDisplayConfig quiverDisplay = new QuiverDisplayConfig();
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/GUIConfig.java]: id argue this belongs into combat category

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/QuiverDisplayConfig.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import io.github.moulberry.moulconfig.annotations.ConfigEditorSlider;
+import io.github.moulberry.moulconfig.annotations.ConfigOption;
+import io.github.moulberry.moulconfig.observer.Property;
+
+public class QuiverDisplayConfig {
+    @Expose
+    @ConfigOption(name = "Enable", desc = "Show the number of arrows you have.")
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean enabled = true;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/QuiverDisplayConfig.java]: either make this default disabled or show "while bow in hand" per default.

File: src/main/java/at/hannibal2/skyhanni/config/features/gui/QuiverDisplayConfig.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Expose
+    @ConfigOption(
+        name = "Reminder After Run",
+        desc = "Reminds you to buy arrows after\n" +
+            "a Dungeons/Kuudra run if you're low."
+    )
+    @ConfigEditorBoolean
+    @FeatureToggle
+    public boolean reminderAfterRun = false;
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/features/gui/QuiverDisplayConfig.java]: id argue to have this default enabled. 
also a bit uncrear: is this part of the display, or can you hide the display always and just have this warning?

File: src/main/java/at/hannibal2/skyhanni/features/gui/QuiverDisplay.kt [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        ) {
+            updateDisplay()
+        }
+    }
+
+    fun isEnabled() = LorenzUtils.inSkyBlock && config.enabled
+
+    @SubscribeEvent
+    fun onConfigFix(event: ConfigUpdaterMigrator.ConfigFixEvent) {
+        event.move(31,"config.inventory.quiverAlert","config.gui.quiverDisplay.lowQuiverNotification")
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/QuiverDisplay.kt]: remove the `config.` part

File: src/main/java/at/hannibal2/skyhanni/features/gui/QuiverDisplay.kt [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        val itemStack = NEUItems.getItemStackOrNull(arrow.internalName.asString()) ?: ItemStack(Items.arrow)
+
+        rarity = itemStack.getItemRarityOrNull()?.chatColorCode ?: "§f"
+        val arrowDisplayName = if (hideAmount || arrow == NONE_ARROW_TYPE) arrow.arrow else StringUtils.pluralize(amount, arrow.arrow)
+
+        if (config.showIcon.get()) {
+            add(Renderable.itemStack(itemStack,1.68))
+        }
+        if (!hideAmount) {
+            add(Renderable.string(" §b${amount}x"))
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/QuiverDisplay.kt]: please add `addSeparators()`

**Review by hannibal002 — CHANGES_REQUESTED** [review]
works fine in game.
maybe split "display" and "warning" up into different categories?

File: src/main/java/at/hannibal2/skyhanni/features/gui/quiver/QuiverDisplay.kt [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    private fun drawDisplay() = buildList {
+        val arrow = arrow ?: return@buildList
+        val itemStack = NEUItems.getItemStackOrNull(arrow.internalName.asString()) ?: ItemStack(Items.arrow)
+
+        val rarity = itemStack.getItemRarityOrNull()?.chatColorCode ?: "§f"
+        val arrowDisplayName =
+            if (hideAmount || arrow == NONE_ARROW_TYPE) arrow.arrow else StringUtils.pluralize(amount, arrow.arrow)
+
+        if (config.showIcon.get()) {
+            add(Renderable.itemStack(itemStack, 1.68))
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/features/gui/quiver/QuiverDisplay.kt]: A bit too large:
image

File: src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.java [hunk 7]
```diff
@@ -68,7 +68,7 @@ public static class ArrowsStorage {
         public String currentArrow = null;
 
         @Expose
-        public Map<NEUInternalName, Float> arrowAmount = new HashMap<>();
+        public Map<NEUInternalName, Integer> arrowAmount = new HashMap<>();
```
Comment (hannibal002) [inline src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.java:71]: this feels it could cause config resets?

**Review by hannibal002 — CHANGES_REQUESTED** [review]
looks like it works fine in game

File: src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.java (hunk 7 above)
Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/config/storage/ProfileSpecificStorage.java:71]: i asked you this some weeks ago and you told me it was fine and didnt need config migration iirc

File: src/main/java/at/hannibal2/skyhanni/features/gui/quiver/QuiverDisplay.kt (hunk 6 above)
Comment (ItsEmpa) [reply src/main/java/at/hannibal2/skyhanni/features/gui/quiver/QuiverDisplay.kt]: need to update it since the itemstack rendering changes

**Review by hannibal002 — APPROVED** [review]

**Pull request merged**